src/inabox/inabox-resources.js

```diff
@@ -0,0 +1,266 @@
+    this.ampdoc_ = ampdoc;
+
+    /** @const {!Window} */
+    this.win = ampdoc.win;
+
+    /** @private @const {!Array<!Resource>} */
+    this.resources_ = [];
+
+    /** @private {number} */
+    this.resourceIdCounter_ = 0;
+
+    /** @private @const {!../service/vsync-impl.Vsync} */
+    this.vsync_ = Services./*OK*/ vsyncFor(this.win);
+
+    /** @const @private {!Pass} */
+    this.pass_ = new Pass(this.win, this.doPass_.bind(this));
+
+    /** @private @const {!Observable} */
+    this.passObservable_ = new Observable();
+
+    /** @const @private {!Deferred} */
+    this.firstPassDone_ = new Deferred();
+  }
+
+  /** @override */
+  get() {
+    return this.resources_.slice(0);
+  }
+
+  /** @override */
+  getAmpdoc() {
+    return this.ampdoc_;
+  }
+
+  /** @override */
+  getMeasuredResources(unusedHostWin, unusedFilterFn) {
+    return Promise.resolve(this.get());
+  }
+
+  /** @override */
+  getResourcesInRect(unusedHostWin, unusedRect, opt_isInPrerender) {
+    return Promise.resolve(this.get());
+  }
+
+  /** @override */
+  getResourceForElement(element) {
+    return Resource.forElement(element);
+  }
+
+  /** @override */
+  getResourceForElementOptional(element) {
+    return Resource.forElementOptional(element);
+  }
+
+  /** @override */
+  getScrollDirection() {
+    return 1;
+  }
+
+  /** @override */
+  add(element) {
+    const resource = new Resource(++this.resourceIdCounter_, element, this);
+    this.resources_.push(resource);
+    dev().fine(TAG, 'resource added:', resource.debugid);
+  }
+
+  /** @override */
+  upgraded(element) {
+    const resource = this.getResourceForElement(element);
+    this.buildThenSchedulePass_(resource);
+    dev().fine(TAG, 'resource upgraded:', resource.debugid);
+  }
+
+  /** @override */
+  remove(element) {
+    const resource = Resource.forElementOptional(element);
+    if (!resource) {
+      return;
+    }
+    const index = this.resources_.indexOf(resource);
+    if (index !== -1) {
+      this.resources_.splice(index, 1);
+    }
+    dev().fine(TAG, 'element removed:', resource.debugid);
+  }
+
+  /** @override */
+  removeForChildWindow(unusedChildWin) {
+    // no child window in inabox
+  }
+
+  /** @override */
+  scheduleLayoutOrPreload(unusedResource) {
+    // all elements are immediately scheduled for layout after being added
+  }
+
+  /** @override */
+  schedulePass(opt_delay) {
+    return this.pass_.schedule(opt_delay);
+  }
+
+  /** @override */
+  onNextPass(callback) {
+    this.passObservable_.add(callback);
+  }
+
+  /** @override */
+  ampInitComplete() {}
+
+  /** @override */
+  updateLayoutPriority(unusedElement, unusedNewLayoutPriority) {
+    // concept of element priority does not exist in inabox
+  }
+
+  /** @override */
+  changeSize(element, newHeight, newWidth, opt_callback, opt_newMargins) {
+    const resource = Resource.forElement(element);
+    resource./*OK*/ changeSize(newHeight, newWidth, opt_newMargins);
```

Comment:
I think this should go into a vsync.mutate callback.

Reply:
done